core/src/main/scalanative/sttp/client4/curl/AbstractCurlBackend.scala

```diff
@@ -25,35 +25,59 @@ import scala.scalanative.unsigned._
 abstract class AbstractCurlBackend[F[_]](_monad: MonadError[F], verbose: Boolean) extends GenericBackend[F, Any] {
   override implicit def monad: MonadError[F] = _monad
 
+  /** Given a [[CurlHandle]], perform the request and return a [[CurlCode]]. */
+  protected def performCurl(c: CurlHandle): F[CurlCode]
+
+  /** Same as [[performCurl]], but also checks and throws runtime exceptions on bad [[CurlCode]]s. */
+  final def perform(c: CurlHandle) = performCurl(c).flatMap(lift)
+
   type R = Any with Effect[F]
 
   override def close(): F[Unit] = monad.unit(())
 
-  private var headers: CurlList = _
-  private var multiPartHeaders: Seq[CurlList] = Seq()
+  /** A request-specific context, with allocated zones and headers. */
+  private class Context() {
+    val zone: Zone = Zone.open()
+    var headers: CurlList = _
```

Comment:
instead of `var`s, mabye we can prepare the headers/multiPartHeaders parameters in `send`, and then create an (immutable) instance of `Context`? E.g. `case class Context(zone: Zone, headers: Option[CurlList], multiPartHeaders: Seq[CurlList])`

Reply:
A bit tricky to do, since `setRequestBody` is setting up the multi part headers as it goes through the multipart body. I can on the other hand just exposes a `addHeader` function to create a header that gets cleared when the context is closed.